spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,216 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.stream.binder.pubsub;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Optional;
+import java.util.StringTokenizer;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.junit.rules.ExternalResource;
+
+/**
+ * Rule for instantiating and tearing down a Pub/Sub emulator instance.
+ */
+public class PubSubEmulator extends ExternalResource {
+	private static final Path EMULATOR_CONFIG_DIR = Paths.get(System.getProperty("user.home")).resolve(
+			Paths.get(".config", "gcloud", "emulators", "pubsub"));
+
+	public static final String ENV_FILE_NAME = "env.yaml";
+
+	private static final Path EMULATOR_CONFIG_PATH = EMULATOR_CONFIG_DIR.resolve(ENV_FILE_NAME);
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubEmulator.class);
+
+	// Binder to Pub/Sub emulator for use in individual tests.
+	private PubSubTestBinder binder;
```

Comment:
This makes this API only tied with the SCSt Binder stuff eliminating more broader usage of such a rule in any our (and end-user) tests where we would like to deal with the Pub/Sub.

Is this really intentional or we can come up with the solution to decouple a Binder environment to another component, e.g. one more JUnit rule?..

Reply:
Good point! I think this rule should only provide the `host` of the emulator that the binder tests can use to create the `PubSubTestBinder`.